### PR TITLE
Refactored validation

### PR DESCRIPTION
**Pull request by jspaaks**
**Related issues**

Refs: #445

(For autoclosure of issues when PR is merged use `Fixes #<issue-number>` syntax)

**Describe the changes made in this pull request**

This PR simplifies the validation of the test files. There is now just one version of the schema that is being tested (i.e. the latest, typically unreleased version, whose changes are not testable with `cffconvert` yet). Also there is just one test, which uses `pytest.parametrize ` to cycle through all `CITATION.cff` files.

The resulting output looks something like this:

```shell
$ pytest tests -v
=========================================== test session starts ===========================================
platform linux -- Python 3.10.12, pytest-7.4.2, pluggy-1.3.0 -- /home/user/github/citation-file-format/citation-file-format/venv/bin/python3
cachedir: .pytest_cache
rootdir: /home/user/github/citation-file-format/citation-file-format
collected 39 items                                                                                        

tests/test_cff_files.py::test_cff_files[pass:reference-conference-paper] PASSED                     [  2%]
tests/test_cff_files.py::test_cff_files[pass:entity-author-with-affiliation] PASSED                 [  5%]
tests/test_cff_files.py::test_cff_files[pass:software-without-a-doi] PASSED                         [  7%]
tests/test_cff_files.py::test_cff_files[pass:bso-toolbox] PASSED                                    [ 10%]
tests/test_cff_files.py::test_cff_files[pass:pcmid8] PASSED                                         [ 12%]
tests/test_cff_files.py::test_cff_files[pass:software-without-a-doi-closed-source] PASSED           [ 15%]
tests/test_cff_files.py::test_cff_files[pass:reference-art] PASSED                                  [ 17%]
tests/test_cff_files.py::test_cff_files[pass:reference-report] PASSED                               [ 20%]
tests/test_cff_files.py::test_cff_files[pass:ls1-mardyn] PASSED                                     [ 23%]
tests/test_cff_files.py::test_cff_files[pass:reference-blog] PASSED                                 [ 25%]
tests/test_cff_files.py::test_cff_files[pass:reference-edited-work] PASSED                          [ 28%]
tests/test_cff_files.py::test_cff_files[pass:nonunique-contributors] PASSED                         [ 30%]
tests/test_cff_files.py::test_cff_files[pass:reference-book] PASSED                                 [ 33%]
tests/test_cff_files.py::test_cff_files[pass:poc] PASSED                                            [ 35%]
tests/test_cff_files.py::test_cff_files[pass:nonunique-authors] PASSED                              [ 38%]
tests/test_cff_files.py::test_cff_files[pass:entity-author-with-ror] PASSED                         [ 41%]
tests/test_cff_files.py::test_cff_files[pass:bjmorgan_bsym] PASSED                                  [ 43%]
tests/test_cff_files.py::test_cff_files[pass:esalmela_haplowinder] PASSED                           [ 46%]
tests/test_cff_files.py::test_cff_files[pass:reference-article] PASSED                              [ 48%]
tests/test_cff_files.py::test_cff_files[pass:xenon-middleware_xenon-adaptors-cloud] PASSED          [ 51%]
tests/test_cff_files.py::test_cff_files[pass:short] PASSED                                          [ 53%]
tests/test_cff_files.py::test_cff_files[pass:nonunique-contact] PASSED                              [ 56%]
tests/test_cff_files.py::test_cff_files[pass:simple] PASSED                                         [ 58%]
tests/test_cff_files.py::test_cff_files[pass:minimal] PASSED                                        [ 61%]
tests/test_cff_files.py::test_cff_files[pass:isbn10] PASSED                                         [ 64%]
tests/test_cff_files.py::test_cff_files[pass:software-executable] PASSED                            [ 66%]
tests/test_cff_files.py::test_cff_files[pass:software-container] PASSED                             [ 69%]
tests/test_cff_files.py::test_cff_files[pass:software-with-a-doi] PASSED                            [ 71%]
tests/test_cff_files.py::test_cff_files[pass:reference-thesis] PASSED                               [ 74%]
tests/test_cff_files.py::test_cff_files[pass:key-complete] PASSED                                   [ 76%]
tests/test_cff_files.py::test_cff_files[pass:software-with-reference] PASSED                        [ 79%]
tests/test_cff_files.py::test_cff_files[pass:software-with-a-doi-expanded] PASSED                   [ 82%]
tests/test_cff_files.py::test_cff_files[fail:invalid-author-array] PASSED                           [ 84%]
tests/test_cff_files.py::test_cff_files[fail:url-trailing-spaces] PASSED                            [ 87%]
tests/test_cff_files.py::test_cff_files[fail:invalid-date] PASSED                                   [ 89%]
tests/test_cff_files.py::test_cff_files[fail:ls1mardyn_ls1mardyn] PASSED                            [ 92%]
tests/test_cff_files.py::test_cff_files[fail:orcid-trailing-chars] PASSED                           [ 94%]
tests/test_cff_files.py::test_cff_files[fail:orcid-trailing-spaces] PASSED                          [ 97%]
tests/test_cff_files.py::test_cff_files[fail:additional-key] PASSED                                 [100%]
```

**Review checklist**

- [x] Please check if the pull request is against the correct branch  
(format/schema/semantic documentation changes: `develop`; typos, meta files, etc.: `main`)
- [x] Please check if all changes are recorded in `CHANGELOG.md` and adapt if necessary.
- [x] Please run tests locally.

```bash
cd $(mktemp -d --tmpdir cff.XXXXXX)
git clone https://github.com/citation-file-format/citation-file-format .
git checkout <branch>
python3 -m venv env
source env/bin/activate
pip install --upgrade pip wheel setuptools
pip install -r requirements.txt
pytest
<do other things>
```

- [ ] `<do other things>`

### CODE REVIEW

**Comment by jspaaks** [comment]
TODOs:

- [ ] We could consider simplifying the validation of cff snippets in schema-guide.md, currently the test tries to iteratively improve the snippet until it passes, a bit difficult to maintain
- [ ] some example files were harvested from the internets, and their origin is reflected in the name. this may come across like we're picking on people (we're not, of course), but perhaps we can rename to something that reflects the type of error
- [ ] I didn't delete the `examples/` directory just yet, but I am in favor of doing so.

**Comment by jspaaks** [comment]
**Update:**

I also added a new way of testing the snippets from `schema-guide.md` using a parameterized test, which I hope is a bit more maintainable. Its output looks like this:

```
tests/test_snippets_from_schema_guide.py::test[L379-L379] PASSED                                   [ 20%]
tests/test_snippets_from_schema_guide.py::test[L382-L382] PASSED                                   [ 20%]
tests/test_snippets_from_schema_guide.py::test[L385-L385] PASSED                                   [ 21%]
tests/test_snippets_from_schema_guide.py::test[L399-L405] PASSED                                   [ 21%]
tests/test_snippets_from_schema_guide.py::test[L415-L435] PASSED                                   [ 21%]
tests/test_snippets_from_schema_guide.py::test[L445-L445] PASSED                                   [ 21%]
tests/test_snippets_from_schema_guide.py::test[L455-L455] PASSED                                   [ 22%]
tests/test_snippets_from_schema_guide.py::test[L465-L465] PASSED                                   [ 22%]
tests/test_snippets_from_schema_guide.py::test[L475-L475] PASSED                                   [ 22%]
tests/test_snippets_from_schema_guide.py::test[L486-L486] PASSED                                   [ 23%]
tests/test_snippets_from_schema_guide.py::test[L489-L489] PASSED                                   [ 23%]
tests/test_snippets_from_schema_guide.py::test[L499-L499] PASSED                                   [ 23%]
tests/test_snippets_from_schema_guide.py::test[L509-L509] PASSED                                   [ 23%]
tests/test_snippets_from_schema_guide.py::test[L512-L512] PASSED                                   [ 24%]
tests/test_snippets_from_schema_guide.py::test[L515-L515] PASSED                                   [ 24%]
tests/test_snippets_from_schema_guide.py::test[L531-L533] XFAIL (deliberately incorrect snippet)   [ 24%]
tests/test_snippets_from_schema_guide.py::test[L536-L539] XFAIL (deliberately incorrect snippet)   [ 25%]
tests/test_snippets_from_schema_guide.py::test[L542-L544] PASSED                                   [ 25%]
tests/test_snippets_from_schema_guide.py::test[L582-L584] PASSED                                   [ 25%]
tests/test_snippets_from_schema_guide.py::test[L594-L596] PASSED                                   [ 25%]
tests/test_snippets_from_schema_guide.py::test[L606-L608] PASSED                                   [ 26%]
tests/test_snippets_from_schema_guide.py::test[L618-L618] PASSED                                   [ 26%]
tests/test_snippets_from_schema_guide.py::test[L621-L621] PASSED                                   [ 26%]
tests/test_snippets_from_schema_guide.py::test[L880-L882] PASSED                                   [ 27%]
tests/test_snippets_from_schema_guide.py::test[L885-L888] FAILED                                   [ 27%]
tests/test_snippets_from_schema_guide.py::test[L898-L898] PASSED                                   [ 27%]
tests/test_snippets_from_schema_guide.py::test[L901-L903] FAILED                                   [ 27%]
tests/test_snippets_from_schema_guide.py::test[L906-L908] FAILED                                   [ 28%]
tests/test_snippets_from_schema_guide.py::test[L911-L913] FAILED                                   [ 28%]
tests/test_snippets_from_schema_guide.py::test[L916-L918] FAILED                                   [ 28%]
tests/test_snippets_from_schema_guide.py::test[L921-L925] FAILED                                   [ 29%]
tests/test_snippets_from_schema_guide.py::test[L935-L935] PASSED                                   [ 29%]
tests/test_snippets_from_schema_guide.py::test[L945-L947] PASSED                                   [ 29%]
```

I mostly like this approach:

- the way of running tests is completely idiomatic, no custom stuff
- the `id` for each snippet corresponds to the line numbers in `schema-guide.md`, which is intuitive
- running just one test  is also possible, e.g. with `pytest tests/test_snippets_from_schema_guide.py::test[L945-L947]`

There are still a whole lot of errors, which are almost exclusively due to missing required properties from `references` examples. I'm planning to do some more work on this.

File: schema-guide.md [hunk 1]
```diff
@@ -885,6 +885,7 @@ authors:
     references:
       - conference:
           country: DE
+          name: The conference name
```
Comment (jspaaks) [inline schema-guide.md:888]: Adding `name` here was the simplest workaround, not sure if we will run into this (i.e. required properties of entities) more in the future.

File: schema-guide.md [hunk 2]
```diff
@@ -922,6 +923,7 @@ authors:
       - conference:
             date-end: "2020-02-02"
             date-start: "2020-01-31"
+            name: The conference name
```
Comment (jspaaks) [inline schema-guide.md:926]: same as other comment

**Review by sdruskat — APPROVED** [review]
Nice one 🎉!

**Comment by sdruskat** [comment]
> TODOs:
> 
>     * [ ]  some example files were harvested from the internets, and their origin is reflected in the name. this may come across like we're picking on people (we're not, of course), but perhaps we can rename to something that reflects the type of error

I can do that in a separate PR.
 
>     * [ ]  I didn't delete the `examples/` directory just yet, but I am in favor of doing so.

Yes.

>     * [ ]  Update _Testing_ section in `README.dev.md`

Yes.

**Pull request merged**